Controllers/AbstractControllers/AbstractAdminController.php

```diff
@@ -0,0 +1,244 @@
+    }
+
+    /**
+     * @param string $logFileType
+     * @param string $logMsg
+     * @param string|null $controller
+     * @param string|null $action
+     * @param string|null $params
+     * @throws \Exception
+     */
+    protected function wrongData(
+        string $logFileType,
+        string $logMsg,
+        string $controller = null,
+        string $action = null,
+        string $params = null
+    ): void {
+        $this->getLogger()->log($logFileType, $logMsg);
+        $this->msgModel->setErrorMsg();
+        $this->prepareRedirect($this->createRedirectString($controller, $action, $params));
+    }
+
+    /**
+     * @param string $uriType
+     * @return AbstractBaseMsgModel
+     * @throws \Exception
+     */
+    protected function getMsgModel(string $uriType = 'default'): AbstractBaseMsgModel
+    {
+        if (!$this->msgModel) {
+            switch ($uriType) {
+                case 'referer' :
+                    return $this->getMsgModelByReferer();
+                case 'request' :
+                    return $this->getMsgModelByRequest();
+                case 'default' :
+                    $this->msgModel = MsgModelsFactory::getMsgModel('default');
+                    return $this->msgModel;
+                default :
+                    throw new \Exception('Unknown MsgModel type in AbstractAdminController :' . " '$uriType'");
+            }
+        } else {
+            return $this->msgModel;
+        }
+    }
+
+    /**
+     * @return AbstractBaseMsgModel
+     * @throws \Exception
+     */
+    protected function getMsgModelByReferer(): AbstractBaseMsgModel
+    {
+        $this->msgModel = MsgModelsFactory::getMsgModel($this->getRefererAdminType(), $this->getRefererAction());
+
+        return $this->msgModel;
+    }
+
+    /**
+     * @return AbstractBaseMsgModel
+     * @throws \Exception
+     */
+    protected function getMsgModelByRequest(): AbstractBaseMsgModel
+    {
+        $this->msgModel = MsgModelsFactory::getMsgModel($this->getRequestAdminType(), $this->getRequestAction());
+
+        return $this->msgModel;
+    }
+
+    /**
+     * @param string $option
+     * @return string
+     * @throws \Exception
+     */
+    protected function getRefererOption(string $option): string
+    {
+        return $this->getServerInfo()->getRefererOption($option);
+    }
+
+    /**
+     * @param string $option
+     * @return string
+     * @throws \Exception
+     */
+    protected function getRequestOption(string $option): string
+    {
+        return $this->getServerInfo()->getRequestOption($option);
+    }
+
+    /**
+     * @return string
+     * @throws \Exception
+     */
+    protected function getRequestAdminType(): string
+    {
+        return $this->getRequestOption(self::ADMIN_TYPE);
+    }
+
+    /**
+     * @return string
+     * @throws \Exception
+     */
+    protected function getRequestController(): string
+    {
+        return $this->getRequestOption(self::CONTROLLER);
```

Comment:
is there a way to avoid lot's of getRequestController getRequestAdminType  getRequestAction

Reply:
I dont exactly understand, what u r talking about.
1) U mean that I need to put all this methods from different controllers under one "roof controller" ?
 OR
2) U mean that i have make something like this getRequestOption(string $option) ;?   And $options will be 'controller',
'action', etc.
Because if u r talking about second option '2)', I told u that i have to spread "string $options" to different methods because of length methods names, for example: 
$this-getRequestAction();
$this-getRequestOption('action');
$this->getServerOption('request', 'action');
As result the simple redirect method will be looks like:
$this->prepareRedirect($this->getRequestController(), $this->getRequestAction());
OR
$this->prepareRedirect($this->getServerOption('request', 'controller'), $this->getServerOption('request', 'action'));
And please dont forget that it well be written not from the beginning of the line, but mostly after 25% of the line